heartbeat autohandler missing

# Bug Report

## System Information

| Type                     | Version/Name |
| ------------------------ | ------------ |
| Operating System Name    | <!-- provide name of OS you're using here, e.g.: Windows, RHEL --> Fedora |
| Operating System Version | <!-- provide the version of the OS you're using here, e.g.: 6.10, 7 --> 27 |
| Python version           | <!-- provide the output of `python -V` you're using --> 2.7.15|
| tlslite-ng version       | <!-- provide the version or commit ID of tlslite-ng version you're using(use `git rev-parse HEAD` in tlslite-ng repo copy you're using or if you're using a packages version, run `python -c 'import tlslite; print(tlslite.__version__)'`, also how it was installed --> 45e81dc4de6b1c074f  |
| tlsfuzzer revision       | <!-- provide the revision of tlsfuzzer that you are using (use `git rev-parse HEAD` in the checked-out tlsfuzzer directory --> fc8fd0128643d|

## Problem description

Running `test-client-compatibility.py` test fails

## Describe how to reproduce the problem

`PYTHONPATH=. python tests/scripts_retention.py tests/tlslite-ng-random-subset.json ../tlslite-1/scripts/tls.py`

## Expected behaviour

test suite passes

## Include errors or backtraces

```
server:stdout:----------------------------------------
test-client-compatibility.py:stdout:Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f92981d8890> (child: None) with last message being: <tlslite.messages.Message object at 0x7f929814f290>
test-client-compatibility.py:stdout:Error while processing
server:stdout:----------------------------------------
server:stdout:Exception happened during processing of request from ('127.0.0.1', 37734)
server:stderr:Traceback (most recent call last):
test-client-compatibility.py:stdout:Traceback (most recent call last):
server:stderr:  File "/usr/lib64/python2.7/SocketServer.py", line 596, in process_request_thread
test-client-compatibility.py:stdout:  File "scripts/test-client-compatibility.py", line 5229, in main
test-client-compatibility.py:stdout:    runner.run()
server:stderr:    self.finish_request(request, client_address)
test-client-compatibility.py:stdout:  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 219, in run
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/integration/tlssocketservermixin.py", line 55, in finish_request
test-client-compatibility.py:stdout:    node.process(self.state, msg)
server:stderr:    if self.handshake(tlsConnection) == True:
test-client-compatibility.py:stdout:  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/expect.py", line 473, in process
server:stderr:  File "../tlslite-1/scripts/tls.py", line 500, in handshake
test-client-compatibility.py:stdout:    self._process_extensions(state, cln_hello, srv_hello)
server:stderr:    sni=sni)
test-client-compatibility.py:stdout:  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/expect.py", line 378, in _process_extensions
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/tlsconnection.py", line 1746, in handshakeServer
test-client-compatibility.py:stdout:    handler = self._get_autohandler(ext_id)
server:stderr:    nextProtos=nextProtos, anon=anon, alpn=alpn, sni=sni):
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/tlsconnection.py", line 1775, in handshakeServerAsync
server:stderr:    for result in self._handshakeWrapperAsync(handshaker, checker):
test-client-compatibility.py:stdout:  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/expect.py", line 348, in _get_autohandler
server:stdout:----------------------------------------
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/tlsconnection.py", line 3567, in _handshakeWrapperAsync
server:stderr:    for result in handshaker:
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/tlsconnection.py", line 1991, in _handshakeServerAsyncHelper
server:stderr:    settings):
test-client-compatibility.py:stdout:    .toStr(ext_id)))
server:stderr:  File "/home/hkario/dev/tlsfuzzer/tlslite/tlsconnection.py", line 3321, in _serverCertKeyExchange
test-client-compatibility.py:stdout:AssertionError: No autohandler for heartbeat
server:stderr:    cipherSuite):
test-client-compatibility.py:stdout:
```

## Additional context

--- a/README.md
+++ b/README.md
@@ -23,7 +23,7 @@ You'll need:
 
  * Python 2.6 or later or Python 3.2 or later
  * [tlslite-ng](https://github.com/tomato42/tlslite-ng)
-   0.8.0-alpha16 or later (note that `tlslite` will *not* work and
+   0.8.0-alpha18 or later (note that `tlslite` will *not* work and
    they conflict with each other)
  * [ecdsa](https://github.com/warner/python-ecdsa)
    python module (dependency of tlslite-ng, should get installed

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,1 +1,1 @@
-tlslite-ng>=0.8.0-alpha16
+tlslite-ng>=0.8.0-alpha18

--- a/tests/test_tlsfuzzer_expect.py
+++ b/tests/test_tlsfuzzer_expect.py
@@ -28,13 +28,14 @@ from tlsfuzzer.expect import Expect, ExpectHandshake, ExpectServerHello, \
         ExpectCertificateVerify, ExpectEncryptedExtensions, \
         ExpectNewSessionTicket, hrr_ext_handler_key_share, \
         hrr_ext_handler_cookie, ExpectHelloRetryRequest, \
-        gen_srv_ext_handler_psk, srv_ext_handler_supp_groups
+        gen_srv_ext_handler_psk, srv_ext_handler_supp_groups, \
+        srv_ext_handler_heartbeat
 
 from tlslite.constants import ContentType, HandshakeType, ExtensionType, \
         AlertLevel, AlertDescription, ClientCertificateType, HashAlgorithm, \
         SignatureAlgorithm, CipherSuite, CertificateType, SSL2HandshakeType, \
         SSL2ErrorDescription, GroupName, CertificateStatusType, ECPointFormat,\
-        SignatureScheme, TLS_1_3_HRR
+        SignatureScheme, TLS_1_3_HRR, HeartbeatMode
 from tlslite.messages import Message, ServerHello, CertificateRequest, \
         ClientHello, Certificate, ServerHello2, ServerFinished, \
         ServerKeyExchange, CertificateStatus, CertificateVerify, \
@@ -44,7 +45,8 @@ from tlslite.extensions import SNIExtension, TLSExtension, \
         NPNExtension, ServerKeyShareExtension, ClientKeyShareExtension, \
         SrvSupportedVersionsExtension, SupportedVersionsExtension, \
         HRRKeyShareExtension, CookieExtension, \
-        SrvPreSharedKeyExtension, PskIdentity, PreSharedKeyExtension
+        SrvPreSharedKeyExtension, PskIdentity, PreSharedKeyExtension, \
+        HeartbeatExtension
 from tlslite.utils.keyfactory import parsePEMKey
 from tlslite.x509certchain import X509CertChain, X509
 from tlslite.extensions import SNIExtension, SignatureAlgorithmsExtension
@@ -362,6 +364,38 @@ class TestServerExtensionProcessors(unittest.TestCase):
             srv_ext_handler_supp_groups(state, ext)
 
         self.assertIn("did not send", str(exc.exception))
+
+    def test_srv_ext_handler_heartbeat_peer_allowed(self):
+        ext = HeartbeatExtension().create(
+            HeartbeatMode.PEER_ALLOWED_TO_SEND)
+        state = None
+
+        srv_ext_handler_heartbeat(state, ext)
+
+    def test_srv_ext_handler_heartbeat_peer_not_allowed(self):
+        ext = HeartbeatExtension().create(
+            HeartbeatMode.PEER_NOT_ALLOWED_TO_SEND)
+        state = None
+
+        srv_ext_handler_heartbeat(state, ext)
+
+    def test_srv_ext_handler_heartbeat_with_empty_ext(self):
+        ext = HeartbeatExtension().create(None)
+        state = None
+
+        with self.assertRaises(AssertionError) as exc:
+            srv_ext_handler_heartbeat(state, ext)
+
+        self.assertIn("Empty mode", str(exc.exception))
+
+    def test_srv_ext_handler_heartbeat_with_invalid_payload(self):
+        ext = HeartbeatExtension().create(3)
+        state = None
+
+        with self.assertRaises(AssertionError) as exc:
+            srv_ext_handler_heartbeat(state, ext)
+
+        self.assertIn("Invalid mode", str(exc.exception))
 
     def test_gen_srv_ext_handler_psk(self):
         psk_settings = [(b'test', b'bad secret'),
@@ -1022,6 +1056,41 @@ class TestExpectServerHello(unittest.TestCase):
         self.assertEqual(state.version, (3, 4))
         self.assertTrue(state.msg_sock.tls13record)
 
+    def test_process_with_tls13_unallowed_extension(self):
+        exp = ExpectServerHello()
+
+        state = ConnectionState()
+        client_hello = ClientHello()
+        client_hello.extensions = []
+        client_hello.cipher_suites = [CipherSuite.TLS_AES_128_GCM_SHA256]
+        ext = SupportedGroupsExtension().create([GroupName.secp256r1])
+        client_hello.extensions.append(ext)
+        c_ks = key_share_gen(GroupName.secp256r1)
+        ext = ClientKeyShareExtension().create([c_ks])
+        client_hello.extensions.append(ext)
+        ext = SupportedVersionsExtension().create([(3, 3), (3, 4)])
+        client_hello.extensions.append(ext)
+        state.handshake_messages.append(client_hello)
+        state.msg_sock = mock.MagicMock()
+
+        s_ext = []
+        s_ks = key_share_gen(GroupName.secp256r1)
+        ext = ServerKeyShareExtension().create(s_ks)
+        s_ext.append(ext)
+        ext = SrvSupportedVersionsExtension().create((3, 4))
+        s_ext.append(ext)
+        ext = SupportedGroupsExtension().create([GroupName.secp256r1])
+        s_ext.append(ext)
+        server_hello = ServerHello().create(version=(3, 3),
+                                            random=bytearray(32),
+                                            session_id=bytearray(0),
+                                            cipher_suite=
+                                            CipherSuite.TLS_AES_128_GCM_SHA256,
+                                            extensions=s_ext)
+
+        with self.assertRaises(AssertionError):
+            exp.process(state, server_hello)
+
 
 class TestExpectServerHelloWithHelloRetryRequest(unittest.TestCase):
     def setUp(self):
@@ -1149,6 +1218,40 @@ class TestExpectHelloRetryRequest(unittest.TestCase):
             exp.process(state, hrr)
 
         self.assertIn("No autohandler for 5119", str(e.exception))
+
+    def test_process_with_tls13_unallowed_extension(self):
+        exp = ExpectHelloRetryRequest()
+
+        state = ConnectionState()
+        client_hello = ClientHello()
+        client_hello.extensions = []
+        client_hello.cipher_suites = [CipherSuite.TLS_AES_128_GCM_SHA256]
+        ext = SupportedGroupsExtension().create([GroupName.secp256r1])
+        client_hello.extensions.append(ext)
+        c_ks = key_share_gen(GroupName.secp256r1)
+        ext = ClientKeyShareExtension().create([c_ks])
+        client_hello.extensions.append(ext)
+        ext = SupportedVersionsExtension().create([(3, 3), (3, 4)])
+        client_hello.extensions.append(ext)
+        state.handshake_messages.append(client_hello)
+        state.msg_sock = mock.MagicMock()
+
+        s_ext = []
+        ext = HRRKeyShareExtension().create(GroupName.secp256r1)
+        s_ext.append(ext)
+        ext = SrvSupportedVersionsExtension().create((3, 4))
+        s_ext.append(ext)
+        ext = SupportedGroupsExtension().create([GroupName.secp256r1])
+        s_ext.append(ext)
+        hrr = ServerHello().create(version=(3, 3),
+                                   random=TLS_1_3_HRR,
+                                   session_id=bytearray(0),
+                                   cipher_suite=
+                                   CipherSuite.TLS_AES_128_GCM_SHA256,
+                                   extensions=s_ext)
+
+        with self.assertRaises(AssertionError):
+            exp.process(state, hrr)
 
 
 class TestExpectServerHello2(unittest.TestCase):

--- a/tlsfuzzer/expect.py
+++ b/tlsfuzzer/expect.py
@@ -14,7 +14,7 @@ import tlslite.utils.tlshashlib as hashlib
 from tlslite.constants import ContentType, HandshakeType, CertificateType,\
         HashAlgorithm, SignatureAlgorithm, ExtensionType,\
         SSL2HandshakeType, CipherSuite, GroupName, AlertDescription, \
-        SignatureScheme, TLS_1_3_HRR
+        SignatureScheme, TLS_1_3_HRR, HeartbeatMode
 from tlslite.messages import ServerHello, Certificate, ServerHelloDone,\
         ChangeCipherSpec, Finished, Alert, CertificateRequest, ServerHello2,\
         ServerKeyExchange, ClientHello, ServerFinished, CertificateStatus, \
@@ -236,6 +236,16 @@ def srv_ext_handler_supp_groups(state, extension):
         raise AssertionError("Server did not send any supported_groups")
 
 
+def srv_ext_handler_heartbeat(state, extension):
+    """Process the heartbeat extension from server."""
+    del state
+    if not extension.mode:
+        raise AssertionError("Empty mode in heartbeat extension.")
+    if extension.mode != HeartbeatMode.PEER_ALLOWED_TO_SEND and \
+       extension.mode != HeartbeatMode.PEER_NOT_ALLOWED_TO_SEND:
+        raise AssertionError("Invalid mode in heartbeat extension.")
+
+
 def _srv_ext_handler_psk(state, extension, psk_configs):
     """Process the pre_shared_key extension from server.
 
@@ -279,7 +289,8 @@ _srv_ext_handler = \
          ExtensionType.ec_point_formats: srv_ext_handler_ec_point,
          ExtensionType.supports_npn: srv_ext_handler_npn,
          ExtensionType.key_share: srv_ext_handler_key_share,
-         ExtensionType.supported_versions: srv_ext_handler_supp_vers}
+         ExtensionType.supported_versions: srv_ext_handler_supp_vers,
+         ExtensionType.heartbeat: srv_ext_handler_heartbeat}
 
 
 _HRR_EXT_HANDLER = \
@@ -290,7 +301,8 @@ _HRR_EXT_HANDLER = \
 _EE_EXT_HANDLER = \
         {ExtensionType.server_name: srv_ext_handler_sni,
          ExtensionType.alpn: srv_ext_handler_alpn,
-         ExtensionType.supported_groups: srv_ext_handler_supp_groups}
+         ExtensionType.supported_groups: srv_ext_handler_supp_groups,
+         ExtensionType.heartbeat: srv_ext_handler_heartbeat}
 
 
 class ExpectServerHello(ExpectHandshake):
@@ -321,12 +333,7 @@ class ExpectServerHello(ExpectHandshake):
         # with what the server sent
         if self.extensions and not srv_hello.extensions:
             raise AssertionError("Server did not send any extensions")
-        elif self.extensions == dict() and srv_hello.extensions:
-            raise AssertionError("Unexpected extension(s) from server "
-                                 "of type: {0}".format(
-                                     ', '.join(ExtensionType.toStr(i.extType)
-                                               for i in srv_hello.extensions)))
-        elif self.extensions and srv_hello.extensions:
+        elif self.extensions is not None and srv_hello.extensions:
             expected = set(self.extensions.keys())
             got = set(i.extType for i in srv_hello.extensions)
             if got != expected:
@@ -336,6 +343,12 @@ class ExpectServerHello(ExpectHandshake):
                                          "{0}".format(
                                              ", ".join((ExtensionType.toStr(i)
                                                         for i in diff))))
+                diff = got.difference(expected)
+                if diff:
+                    raise AssertionError("Server sent unexpected extension(s):"
+                                         " {0}".format(
+                                             ", ".join(ExtensionType.toStr(i)
+                                                       for i in diff)))
 
     @staticmethod
     def _get_autohandler(ext_id):
@@ -349,8 +362,23 @@ class ExpectServerHello(ExpectHandshake):
 
     def _process_extensions(self, state, cln_hello, srv_hello):
         """Check if extensions are correct."""
+        sh_supported = [ExtensionType.pre_shared_key,
+                        ExtensionType.supported_versions,
+                        ExtensionType.key_share]
+        hrr_supported = [ExtensionType.cookie,
+                         ExtensionType.supported_versions,
+                         ExtensionType.key_share]
         for ext in srv_hello.extensions:
             ext_id = ext.extType
+            if state.version > (3, 3) and \
+                    ((srv_hello.random != TLS_1_3_HRR and
+                      ext_id not in sh_supported) or
+                     (srv_hello.random == TLS_1_3_HRR and
+                      ext_id not in hrr_supported)):
+                raise AssertionError("Server sent unallowed "
+                                     "extension of type {0}"
+                                     .format(ExtensionType
+                                             .toStr(ext_id)))
             cl_ext = cln_hello.getExtension(ext_id)
             if ext_id == ExtensionType.renegotiation_info and \
                     CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV \
@@ -366,13 +394,8 @@ class ExpectServerHello(ExpectHandshake):
                                              .toStr(ext_id)))
             handler = None
             if self.extensions:
-                try:
-                    handler = self.extensions[ext_id]
-                except KeyError:
-                    raise AssertionError("Unexpected extension from "
-                                         "server of type {0}"
-                                         .format(ExtensionType
-                                                 .toStr(ext_id)))
+                handler = self.extensions[ext_id]
+
             # use automatic handlers for some extensions
             if handler is None:
                 handler = self._get_autohandler(ext_id)
@@ -469,10 +492,10 @@ class ExpectServerHello(ExpectHandshake):
         state.extended_master_secret = False
         state.encrypt_then_mac = False
 
+        self._compare_extensions(srv_hello)
+
         if srv_hello.extensions:
             self._process_extensions(state, cln_hello, srv_hello)
-
-        self._compare_extensions(srv_hello)
 
         if state.version > (3, 3):
             self._setup_tls13_handshake_keys(state)
@@ -1083,7 +1106,7 @@ class ExpectEncryptedExtensions(ExpectHandshake):
                         1,  # max_fragment_length - RFC 6066
                         ExtensionType.supported_groups,
                         14,  # use_srtp - RFC 5764
-                        15,  # heartbeat - RFC 6520
+                        ExtensionType.heartbeat,  # RFC 6520
                         ExtensionType.alpn,
                         19,  # client_certificate_type
                              # draft-ietf-tls-tls13-28 / RFC 7250

